### PR TITLE
Solution for mirror words challenge

### PR DESCRIPTION
**Pull request by Aseel-AbuKmail**
Name: Mirror Words Challenge
About: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by abdoalsir** [comment]
Hi Aseel,
I hope you are having a great day.
I see that your solutions didn't pass some checks, if you need assistance feel free to reach out

**Comment by Aseel-AbuKmail** [comment]
> Hi Aseel, I hope you are having a great day. I see that your solutions didn't pass some checks, if you need assistance feel free to reach out

Hello, Abdulrahman!
Thank you for offering to help; I really appreciate it!

File: solutions/tests/test_mirror_words_challenge.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+# Add the directory containing mirror_words_challenge.py to the module search path
+sys.path.append(
+    "C:/Users/pc/.vscode/VS code Files/MIT - Project/ET6-foundations-group-17-main/solutions"
+)
+
+# Add the project root directory to sys.path
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../../")))
+
+# from mirror_words_challenge import reverse_words
+# from solutions.mirror_words_challenge import reverse_words
```
Comment (abdoalsir) [inline solutions/tests/test_mirror_words_challenge.py:29]: If these lines are unnecessary, delete them. However, if they are intentionally retained for reference or debugging purposes, add a clarifying comment explaining why they are there.

File: solutions/mirror_words_challenge.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# mirror_words_challenge
+import re
+
+
+def reverse_words(sentence):
+    """
+    Reverses each word in a given sentence while maintaining the order of the words
+    and correctly handling punctuation marks at the end of words.
+
```
Comment (abdoalsir) [inline solutions/mirror_words_challenge.py:9]: The docstring is well written and it can be even better with a strategy to make the code even easier to understand for others who might work with or review it in the future.

File: solutions/tests/test_mirror_words_challenge.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+# test_mirror_words_challenge
```
Comment (abdoalsir) [inline solutions/tests/test_mirror_words_challenge.py:1]: The file name already shows what file it's, so I think it would be better to remove this comment so the code appears more clean and professional.

File: solutions/mirror_words_challenge.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# mirror_words_challenge
```
Comment (abdoalsir) [inline solutions/mirror_words_challenge.py:1]: The file name already shows what file it's, so I think it would be better to remove this comment so the code appears more clean and professional

**Review by abdoalsir — COMMENTED** [review]
# Overall Feedback:  
The function and test files are well-constructed and fulfill many criteria for clear and maintainable code. You have effectively utilized docstrings and tests, and the function successfully passes all provided checks. Nevertheless, there are several areas where enhancements could be made to increase the robustness and user-friendliness of the code:
- Function Naming: The name of the function could be more descriptive to eliminate any ambiguity regarding its purpose. Aligning it with the file name (e.g., mirror_words) may clarify its functionality.
- Type Annotations: Incorporating type annotations for both the input argument and the return value would enhance clarity for readers who may not refer to the docstring.
- Docstring Enhancements: Although your docstring is clear, including information about the treatment of special characters and numbers (as indicated in your tests) along with a high-level overview of the strategy would improve its thoroughness and usefulness for others who may review or extend your code.
- Doctests: Adding a third doctest that illustrates the handling of special characters would further enrich the documentation of the function.
- Code Structure: Enhancing the spacing and organization of the code will contribute to making the implementation more readable and accessible to a wider audience.
- Commented Lines: If the commented-out import statements are unnecessary, consider removing them. If they serve a purpose, please add a clarifying comment to explain their intent.

Overall, excellent work!😇 Implementing these recommendations will raise the quality of your code from great to outstanding.

**Comment by Aseel-AbuKmail** [comment]
@abdoalsir  
Hi, I’ve implemented your suggested feedback. If you don’t have any further suggestions, please go ahead and merge it into the main branch and delete my challenge branch.

**Comment by Aseel-AbuKmail** [comment]
**Lessons Learned** from My Mistakes

How to Resolve ET CI Check Issues:

**1. For py_formatting:**
Run the following commands in the terminal:

- pip install ruff
- ruff format
- ruff check

OR:

Run the following commands in the terminal:
- pip install black
- black --check <file_location>
- black <file_location> 

**2. For py_test:**
- Verify that every file is in the correct path.
